libyul/backends/evm/StackLayoutGenerator.cpp

```diff
@@ -0,0 +1,560 @@
+		size_t targetSize() { return post.size(); }
+		void pop() { layout.pop_back(); }
+		void pushOrDupTarget(size_t _offset) { layout.push_back(post.at(_offset)); }
+	};
+
+	Shuffler<ShuffleOperations>::shuffle(_layout, _post, _generateSlotOnTheFly);
+
+	// Now we can construct the ideal layout before the operation.
+	// "layout" has the declared variables in the desired position and
+	// for any PreviousSlot{x}, x yields the ideal place of the slot before the declaration.
+	vector<optional<StackSlot>> idealLayout(_post.size(), nullopt);
+	for (auto const& [slot, idealPosition]: ranges::zip_view(_post, _layout))
+		if (PreviousSlot* previousSlot = std::get_if<PreviousSlot>(&idealPosition))
+			idealLayout.at(previousSlot->slot) = slot;
+
+	while (!idealLayout.empty() && !idealLayout.back())
+		idealLayout.pop_back();
+
+	return idealLayout | ranges::views::transform([](optional<StackSlot> s) {
+		yulAssert(s, "");
+		return *s;
+	}) | ranges::to<Stack>;
+}
+}
+
+Stack StackLayoutGenerator::propagateStackThroughOperation(Stack _exitStack, CFG::Operation const& _operation)
+{
+	Stack& stack = _exitStack;
+
+	// This is a huge tradeoff between code size, gas cost and stack size.
+	auto generateSlotOnTheFly = [&](StackSlot const&) {
+		//return stack.size() > 12 && canBeFreelyGenerated(_slot);
+		// return canBeFreelyGenerated(_slot);
+		return false;
+	};
+
+	size_t previousLayoutSize = stack.size();
+	for (auto const& slot: stack)
+		if (util::findOffset(_operation.output, slot) || generateSlotOnTheFly(slot))
+				--previousLayoutSize;
+
+	auto layout = ranges::views::iota(0u, previousLayoutSize) |
+		ranges::views::transform([](size_t _index) { return PreviousSlot{_index}; }) |
+		ranges::to<vector<variant<PreviousSlot, StackSlot>>>;
+	// The call produces a known sequence of values.
+	layout += _operation.output;
+
+	stack = createIdealLayout(stack, layout, generateSlotOnTheFly);
+
+	if (auto const* assignment = get_if<CFG::Assignment>(&_operation.operation))
+		for (auto& stackSlot: stack)
+			if (auto const* varSlot = get_if<VariableSlot>(&stackSlot))
+				yulAssert(!util::findOffset(assignment->variables, *varSlot), "");
+
+	for (StackSlot const& input: _operation.input)
+		stack.emplace_back(input);
+
+	m_layout.operationEntryLayout[&_operation] = stack;
+
+	// Remove anything from the stack top that can be freely generated or dupped from deeper on the stack.
+	while (!stack.empty())
+	{
+		if (canBeFreelyGenerated(stack.back()))
+			stack.pop_back();
+		else if (auto offset = util::findOffset(stack | ranges::views::reverse | ranges::views::drop(1), stack.back()))
+		{
+			if (*offset + 2 < 16)
+				stack.pop_back();
+			else
+				break;
+		}
+		else
+			break;
+	}
+
+	// TODO: there may be a better criterion than overall stack size.
+	if (stack.size() > 12)
+		// Deduplicate and remove slots that can be freely generated.
+		stack = compressStack(move(stack));
+	return stack;
+}
+
+Stack StackLayoutGenerator::compressStack(Stack _stack)
```

Comment:
Any reason for sorting the functions in a different order than in the header?

Reply:
Yes, the gcc code generator will visit the functions in different order and their optimizer works better in this precise order and it gives 20% speedup, I checked :-)!

[I'll fix the order]

Reply:
> Yes, the gcc code generator will visit the functions in different order and their optimizer works better in this precise order and it gives 20% speedup, I checked :-)!

How are you computing this?

Reply:
I was kidding - I just changed the order.